Mathlib/Data/Polynomial/Smeval.lean

```diff
@@ -113,37 +128,169 @@ def smeval.linearMap : R[X] →ₗ[R] S where
+  | n + 1 => by
+    rw [add_comm, npow_add, mul_assoc, npow_one, smeval_X_mul, smeval_X_pow_mul n, npow_add,
+      smeval_X_pow_assoc, npow_one]
+
+theorem smeval_C_mul : (C r * p).smeval x = r • p.smeval x := by
+  induction p using Polynomial.induction_on' with
+  | h_add p q ph qh =>
+    simp only [mul_add, smeval_add, ph, qh, smul_add]
+  | h_monomial n b =>
+    simp only [C_mul_monomial, smeval_monomial, mul_smul]
+
+theorem smeval_monomial_mul (n : ℕ) :
+    (monomial n r * p).smeval x = r • (x ^ n * p.smeval x) := by
+  induction p using Polynomial.induction_on' with
+  | h_add r s hr hs =>
+    simp only [add_comp, hr, hs, smeval_add, add_mul]
+    rw [← C_mul_X_pow_eq_monomial, mul_assoc, smeval_C_mul, smeval_X_pow_mul, smeval_add]
+  | h_monomial n a =>
+    rw [smeval_monomial, monomial_mul_monomial, smeval_monomial, npow_add, mul_smul, mul_smul_comm]
+
+theorem smeval_mul : (p * q).smeval x  = p.smeval x * q.smeval x := by
+  induction p using Polynomial.induction_on' with
+  | h_add r s hr hs =>
+    simp only [add_comp, hr, hs, smeval_add, add_mul]
+  | h_monomial n a =>
+    simp only [smeval_monomial, smeval_C_mul, smeval_mul_X_pow, smeval_monomial_mul, smul_mul_assoc]
+
+theorem smeval_pow : ∀ (n : ℕ), (p^n).smeval x = (p.smeval x)^n
+  | 0 => by
+    simp only [npow_zero, smeval_one, one_smul]
+  | n + 1 => by
+    rw [npow_add, smeval_mul, smeval_pow n, pow_one, npow_add, npow_one]
+
+theorem smeval_comp : (p.comp q).smeval x  = p.smeval (q.smeval x) := by
+  induction p using Polynomial.induction_on' with
+  | h_add r s hr hs =>
+    simp [add_comp, hr, hs, smeval_add]
+  | h_monomial n a =>
+    simp [smeval_monomial, smeval_C_mul, smeval_pow]
+
+end NatPowAssoc
+
+section Algebra
 
 theorem aeval_coe_eq_smeval {R : Type*} [CommSemiring R] {S : Type*} [Semiring S] [Algebra R S]
-    (x : S) : ⇑(aeval x) = fun p => @smeval S _ _ x R _ _ p := by
+    (x : S) : ⇑(aeval x) = fun (p : R[X]) => p.smeval x := by
   refine funext ?_
   intro
   rw [aeval_def, eval₂_def, Algebra.algebraMap_eq_smul_one', smeval_def]
   simp only [Algebra.smul_mul_assoc, one_mul]
   exact rfl
 
-end Comparisons
+/-- `Polynomial.smeval` as an algebra map. -/
+def smeval.algebraMap (R : Type*) [CommSemiring R] {S : Type*} [Semiring S] [Algebra R S]
+    (x : S) : R[X] →ₐ[R] S where
+  toFun := fun p => p.smeval x
+  map_one' := by simp only [smeval_one, pow_zero, one_smul]
+  map_mul' p q := smeval_mul R p q x
+  map_zero' := by simp only [smeval_zero]
+  map_add' := fun p q => smeval_add R p q x
+  commutes' r := by
+    dsimp only
+    rw [← C_eq_algebraMap, Algebra.algebraMap_eq_smul_one, smeval_C, pow_zero]
+
+theorem smeval.algebraMap.apply {R : Type*} [CommSemiring R] {S : Type*} [Semiring S] [Algebra R S]
+    (x : S) (p : R[X]) : smeval.algebraMap R x p = p.smeval x := rfl
+
+theorem aeval_eq_smeval {R : Type*} [CommSemiring R] {S : Type*} [Semiring S] [Algebra R S]
+    (x : S) : Polynomial.aeval x = smeval.algebraMap R x := by
```

Comment:
Given these are equal, do we actually need `smeval.algebraMap`? I think we could get away with just showing that
```suggestion
    (x : S) (p : R[X]) : Polynomial.aeval x p = p.smeval x := by
```

Reply:
I have removed `smeval.algebraMap`.